tensorflow/python/summary/summary_iterator.py

```diff
@@ -301,26 +306,35 @@ def __init__(self, queue, ev_writer, flush_secs):
     """
     threading.Thread.__init__(self)
     self.daemon = True
+    self._stop_flag = False
     self._queue = queue
     self._ev_writer = ev_writer
     self._flush_secs = flush_secs
     # The first event will be flushed immediately.
     self._next_event_flush_time = 0
+    self._fetch_timeout = 1
 
   def run(self):
     while True:
-      event = self._queue.get()
+      event = self._queue.get(timeout=self._fetch_timeout)
       try:
-        self._ev_writer.WriteEvent(event)
+        if event:
+          self._ev_writer.WriteEvent(event)
         # Flush the event writer every so often.
         now = time.time()
-        if now > self._next_event_flush_time:
+        if now > self._next_event_flush_time or self._stop_flag:
           self._ev_writer.Flush()
           # Do it again in two minutes.
           self._next_event_flush_time = now + self._flush_secs
       finally:
         self._queue.task_done()
 
+      if self._stop_flag:
+        break
```

Comment:
Thank a lot for nice analyze, actually I attempt to drop messages before because I would like to exit immediately. But I agree with all your ideas, let me change to make sure all events flush before exit. Thank you very much for your inputs!